File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/BundleRewriteArtifacts.java
package com.tongsr.kaleido.gradle;

import com.google.protobuf.CodedInputStream;
import com.google.protobuf.CodedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Comparator;
import java.util.List;

final class BundleRewriteArtifacts {
    static final String PLAN_SCHEMA = "BundleRewritePlan.v1";
    static final String RECEIPT_SCHEMA = "TransformReceipt.v1";
    static final String PRODUCER = "KaleidoBundleRewrite/1";

    private BundleRewriteArtifacts() {}

    static byte[] encodePlan(Plan plan) throws IOException {
        return message(output -> {
            output.writeString(1, plan.schema());
            output.writeString(2, plan.producer());
            output.writeString(3, plan.project());
            output.writeString(4, plan.variant());
            output.writeString(5, plan.inputAabSha256());
            output.writeString(6, plan.resourceTableSha256());
            for (var resource : plan.resources()) output.writeByteArray(7, message(nested -> {
                nested.writeUInt32(1, resource.id());
                nested.writeString(2, resource.packageName());
                nested.writeString(3, resource.type());
                nested.writeString(4, resource.originalName());
                nested.writeString(5, resource.targetName());
                nested.writeString(6, resource.action());
                nested.writeString(7, resource.reason());
                for (var path : resource.originalPaths()) nested.writeString(8, path);
                for (var path : resource.targetPaths()) nested.writeString(9, path);
                nested.writeBool(10, resource.nameProtected());
                nested.writeBool(11, resource.pathProtected());
            }));
            for (var entry : plan.entries()) output.writeByteArray(8, message(nested -> {
                nested.writeString(1, entry.inputPath());
                nested.writeString(2, entry.inputSha256());
                nested.writeUInt32(3, entry.compressionMethod());
                nested.writeString(4, entry.outputPath());
                nested.writeString(5, entry.action());
                nested.writeBool(6, entry.preservePayload());
            }));
            for (var outputPath : plan.expectedOutputs()) output.writeString(9, outputPath);
            for (var reference : plan.references()) output.writeByteArray(10, message(nested -> {
                nested.writeString(1, reference.origin());
                nested.writeString(2, reference.fieldPath());
                nested.writeString(3, reference.kind());
                nested.writeUInt32(4, reference.resourceId());
                nested.writeString(5, reference.originalValue());
                nested.writeString(6, reference.targetValue());
            }));
            for (var control : plan.controls()) output.writeByteArray(12, message(nested -> {
                nested.writeString(1, control.kind());
                nested.writeString(2, control.target());
                nested.writeString(3, control.inputSha256());
                nested.writeString(4, control.outputSha256());
                nested.writeString(5, control.action());
                nested.writeString(6, control.reason());
                nested.writeBool(7, control.protectedTarget());
            }));
        });
    }

    static Plan decodePlan(byte[] bytes, String project, String variant) throws IOException {
        var input = CodedInputStream.newInstance(bytes);
        var values = new String[] {"", "", "", "", "", ""};
        var resources = new java.util.ArrayList<ResourceDecision>();
        var entries = new java.util.ArrayList<EntryDecision>();
        var outputs = new java.util.ArrayList<String>();
        var references = new java.util.ArrayList<ReferenceDecision>();
        var controls = new java.util.ArrayList<ControlDecision>();
        while (!input.isAtEnd()) {
            var tag = input.readTag();
            switch (tag) {
                case 10, 18, 26, 34, 42, 50 ->
                        values[(tag - 10) / 8] = input.readStringRequireUtf8();
                case 58 -> resources.add(decodeResource(input.readByteArray()));
                case 66 -> entries.add(decodeEntry(input.readByteArray()));
                case 74 -> outputs.add(input.readStringRequireUtf8());
                case 82 -> references.add(decodeReference(input.readByteArray()));
                case 98 -> controls.add(decodeControl(input.readByteArray()));
                case 0 -> { }
                default -> input.skipField(tag);
            }
        }
        if (!PLAN_SCHEMA.equals(values[0])) {
            throw failure(project, variant, "schema",
                    "Unknown Bundle Rewrite Plan major: " + values[0],
                    "Regenerate the plan with this Kaleido version");
        }
        return new Plan(values[0], values[1], values[2], values[3], values[4], values[5],
                resources, entries, outputs, references, controls);
    }

    static byte[] encodeReceipt(Receipt receipt) throws IOException {
        return message(output -> {
            output.writeString(1, receipt.schema());
            output.writeString(2, receipt.producer());
            output.writeString(3, receipt.project());
            output.writeString(4, receipt.variant());
            output.writeString(5, receipt.planSha256());
            output.writeString(6, receipt.inputAabSha256());
            output.writeString(7, receipt.outputAabSha256());
            output.writeString(8, receipt.resourceMappingSha256());
            output.writeUInt32(9, receipt.resourceIdCount());
            output.writeUInt32(10, receipt.renamedResourceCount());
            output.writeUInt32(11, receipt.renamedPathCount());
            output.writeBool(12, receipt.referenceClosureValidated());
            output.writeBool(13, receipt.preservedPayloadsValidated());
            output.writeBool(14, receipt.bundletoolValidated());
            output.writeUInt32(15, receipt.referenceCount());
            for (var digest : receipt.preservedPayloadDigests()) output.writeString(16, digest);
            for (var digest : receipt.protectionDigests()) output.writeString(17, digest);
        });
    }

    private static ResourceDecision decodeResource(byte[] bytes) throws IOException {
        var input = CodedInputStream.newInstance(bytes);
        var id = 0;
        var values = new String[] {"", "", "", "", "", ""};
        var originalPaths = new java.util.ArrayList<String>();
        var targetPaths = new java.util.ArrayList<String>();
        var nameProtected = false;
        var pathProtected = false;
        while (!input.isAtEnd()) {
            var tag = input.readTag();
            switch (tag) {
                case 8 -> id = input.readUInt32();
                case 18, 26, 34, 42, 50, 58 ->
                        values[(tag - 18) / 8] = input.readStringRequireUtf8();
                case 66 -> originalPaths.add(input.readStringRequireUtf8());
                case 74 -> targetPaths.add(input.readStringRequireUtf8());
                case 80 -> nameProtected = input.readBool();
                case 88 -> pathProtected = input.readBool();
                case 0 -> { }
                default -> input.skipField(tag);
            }
        }
        return new ResourceDecision(id, values[0], values[1], values[2], values[3],
                values[4], values[5], originalPaths, targetPaths,
                nameProtected, pathProtected);
    }

    private static EntryDecision decodeEntry(byte[] bytes) throws IOException {
        var input = CodedInputStream.newInstance(bytes);
        var values = new String[] {"", "", "", ""};
        var method = 0;
        var preserve = false;
        while (!input.isAtEnd()) {
            var tag = input.readTag();
            switch (tag) {
                case 10 -> values[0] = input.readStringRequireUtf8();
                case 18 -> values[1] = input.readStringRequireUtf8();
                case 24 -> method = input.readUInt32();
                case 34 -> values[2] = input.readStringRequireUtf8();
                case 42 -> values[3] = input.readStringRequireUtf8();
                case 48 -> preserve = input.readBool();
                case 0 -> { }
                default -> input.skipField(tag);
            }
        }
        return new EntryDecision(values[0], values[1], method, values[2], values[3], preserve);
    }

    private static ReferenceDecision decodeReference(byte[] bytes) throws IOException {
        var input = CodedInputStream.newInstance(bytes);
        var values = new String[] {"", "", "", "", ""};
        var id = 0;
        while (!input.isAtEnd()) {
            var tag = input.readTag();
            switch (tag) {
                case 10, 18, 26 -> values[(tag - 10) / 8] = input.readStringRequireUtf8();
                case 32 -> id = input.readUInt32();
                case 42 -> values[3] = input.readStringRequireUtf8();
                case 50 -> values[4] = input.readStringRequireUtf8();
                case 0 -> { }
                default -> input.skipField(tag);
            }
        }
        return new ReferenceDecision(values[0], values[1], values[2], id,
                values[3], values[4]);
    }

    private static ControlDecision decodeControl(byte[] bytes) throws IOException {
        var input = CodedInputStream.newInstance(bytes);
        var values = new String[] {"", "", "", "", "", ""};
        var protectedTarget = false;
        while (!input.isAtEnd()) {
            var tag = input.readTag();
            switch (tag) {
                case 10, 18, 26, 34, 42, 50 ->
                        values[(tag - 10) / 8] = input.readStringRequireUtf8();
                case 56 -> protectedTarget = input.readBool();
                case 0 -> { }
                default -> input.skipField(tag);
            }
        }
        return new ControlDecision(values[0], values[1], values[2], values[3],
                values[4], values[5], protectedTarget);
    }

    private static byte[] message(Writer writer) throws IOException {
        var bytes = new ByteArrayOutputStream();
        var output = CodedOutputStream.newInstance(bytes);
        writer.write(output);
        output.flush();
        return bytes.toByteArray();
    }

    private static org.gradle.api.GradleException failure(
            String project, String variant, String target, String reason, String repair) {
        return new KaleidoDiagnostic("KLD-BUNDLE-001", project, variant, "bundle-rewrite",
                PLAN_SCHEMA, target, reason, repair).failure();
    }

    @FunctionalInterface private interface Writer {
        void write(CodedOutputStream output) throws IOException;
    }

    record Plan(String schema, String producer, String project, String variant,
                String inputAabSha256, String resourceTableSha256,
                List<ResourceDecision> resources, List<EntryDecision> entries,
                List<String> expectedOutputs, List<ReferenceDecision> references,
                List<ControlDecision> controls) {
        Plan {
            resources = resources.stream().sorted((left, right) ->
                            Integer.compareUnsigned(left.id(), right.id()))
                    .toList();
            entries = entries.stream().sorted(Comparator.comparing(EntryDecision::inputPath))
                    .toList();
            expectedOutputs = expectedOutputs.stream().sorted().toList();
            references = references.stream().sorted(Comparator
                    .comparing(ReferenceDecision::origin)
                    .thenComparing(ReferenceDecision::fieldPath)
                    .thenComparing(ReferenceDecision::kind)).toList();
            controls = controls.stream().sorted(Comparator
                    .comparing(ControlDecision::kind)
                    .thenComparing(ControlDecision::target)).toList();
        }
    }

    record ResourceDecision(int id, String packageName, String type, String originalName,
                            String targetName, String action, String reason,
                            List<String> originalPaths, List<String> targetPaths,
                            boolean nameProtected, boolean pathProtected) {
        ResourceDecision {
            if (originalPaths.size() != targetPaths.size()) {
                throw new IllegalArgumentException("Resource path mapping is not one-to-one");
            }
            var order = java.util.stream.IntStream.range(0, originalPaths.size()).boxed()
                    .sorted(Comparator.comparing(originalPaths::get)).toList();
            var canonicalOriginals = order.stream().map(originalPaths::get).toList();
            var canonicalTargets = order.stream().map(targetPaths::get).toList();
            originalPaths = canonicalOriginals;
            targetPaths = canonicalTargets;
        }
    }

    record EntryDecision(String inputPath, String inputSha256, int compressionMethod,
                         String outputPath, String action, boolean preservePayload) {}

    record ReferenceDecision(String origin, String fieldPath, String kind, int resourceId,
                             String originalValue, String targetValue) {}

    record ControlDecision(String kind, String target, String inputSha256,
                           String outputSha256, String action, String reason,
                           boolean protectedTarget) {}

    record Receipt(String schema, String producer, String project, String variant,
                   String planSha256, String inputAabSha256, String outputAabSha256,
                   String resourceMappingSha256, int resourceIdCount,
                   int renamedResourceCount, int renamedPathCount,
                   boolean referenceClosureValidated, boolean preservedPayloadsValidated,
                   boolean bundletoolValidated, int referenceCount,
                   List<String> preservedPayloadDigests, List<String> protectionDigests) {
        Receipt {
            preservedPayloadDigests = preservedPayloadDigests.stream().sorted().toList();
            protectionDigests = protectionDigests.stream().sorted().toList();
        }
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/BundleRewriteArtifacts.kt
package com.tongsr.kaleido.gradle

import com.google.protobuf.CodedInputStream
import com.google.protobuf.CodedOutputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import org.gradle.api.GradleException

internal object BundleRewriteArtifacts {
    const val PLAN_SCHEMA: String = "BundleRewritePlan.v1"
    const val RECEIPT_SCHEMA: String = "TransformReceipt.v1"
    const val PRODUCER: String = "KaleidoBundleRewrite/1"

    @JvmStatic
    @Throws(IOException::class)
    fun encodePlan(plan: Plan): ByteArray {
        val resources = plan.resources.sortedWith { left, right ->
            Integer.compareUnsigned(left.id, right.id)
        }
        val entries = plan.entries.sortedBy(EntryDecision::inputPath)
        val expectedOutputs = plan.expectedOutputs.sorted()
        val references = plan.references.sortedWith(
            compareBy(ReferenceDecision::origin)
                .thenBy(ReferenceDecision::fieldPath)
                .thenBy(ReferenceDecision::kind),
        )
        val controls = plan.controls.sortedWith(
            compareBy(ControlDecision::kind).thenBy(ControlDecision::target),
        )
        return message { output ->
            output.writeString(1, plan.schema)
            output.writeString(2, plan.producer)
            output.writeString(3, plan.project)
            output.writeString(4, plan.variant)
            output.writeString(5, plan.inputAabSha256)
            output.writeString(6, plan.resourceTableSha256)
            for (resource in resources) {
                output.writeByteArray(7, message { nested ->
                    nested.writeUInt32(1, resource.id)
                    nested.writeString(2, resource.packageName)
                    nested.writeString(3, resource.type)
                    nested.writeString(4, resource.originalName)
                    nested.writeString(5, resource.targetName)
                    nested.writeString(6, resource.action)
                    nested.writeString(7, resource.reason)
                    for (path in resource.originalPaths) nested.writeString(8, path)
                    for (path in resource.targetPaths) nested.writeString(9, path)
                    nested.writeBool(10, resource.nameProtected)
                    nested.writeBool(11, resource.pathProtected)
                })
            }
            for (entry in entries) {
                output.writeByteArray(8, message { nested ->
                    nested.writeString(1, entry.inputPath)
                    nested.writeString(2, entry.inputSha256)
                    nested.writeUInt32(3, entry.compressionMethod)
                    nested.writeString(4, entry.outputPath)
                    nested.writeString(5, entry.action)
                    nested.writeBool(6, entry.preservePayload)
                })
            }
            for (outputPath in expectedOutputs) output.writeString(9, outputPath)
            for (reference in references) {
                output.writeByteArray(10, message { nested ->
                    nested.writeString(1, reference.origin)
                    nested.writeString(2, reference.fieldPath)
                    nested.writeString(3, reference.kind)
                    nested.writeUInt32(4, reference.resourceId)
                    nested.writeString(5, reference.originalValue)
                    nested.writeString(6, reference.targetValue)
                })
            }
            for (control in controls) {
                output.writeByteArray(12, message { nested ->
                    nested.writeString(1, control.kind)
                    nested.writeString(2, control.target)
                    nested.writeString(3, control.inputSha256)
                    nested.writeString(4, control.outputSha256)
                    nested.writeString(5, control.action)
                    nested.writeString(6, control.reason)
                    nested.writeBool(7, control.protectedTarget)
                })
            }
        }
    }

    @JvmStatic
    @Throws(IOException::class)
    fun decodePlan(bytes: ByteArray, project: String, variant: String): Plan {
        val input = CodedInputStream.newInstance(bytes)
        val values = arrayOf("", "", "", "", "", "")
        val resources = ArrayList<ResourceDecision>()
        val entries = ArrayList<EntryDecision>()
        val outputs = ArrayList<String>()
        val references = ArrayList<ReferenceDecision>()
        val controls = ArrayList<ControlDecision>()
        while (!input.isAtEnd) {
            val tag = input.readTag()
            when (tag) {
                10, 18, 26, 34, 42, 50 ->
                    values[(tag - 10) / 8] = input.readStringRequireUtf8()
                58 -> resources.add(decodeResource(input.readByteArray()))
                66 -> entries.add(decodeEntry(input.readByteArray()))
                74 -> outputs.add(input.readStringRequireUtf8())
                82 -> references.add(decodeReference(input.readByteArray()))
                98 -> controls.add(decodeControl(input.readByteArray()))
                0 -> { }
                else -> input.skipField(tag)
            }
        }
        if (PLAN_SCHEMA != values[0]) {
            throw failure(
                project,
                variant,
                "schema",
                "Unknown Bundle Rewrite Plan major: " + values[0],
                "Regenerate the plan with this Kaleido version",
            )
        }
        return Plan(
            values[0], values[1], values[2], values[3], values[4], values[5],
            resources, entries, outputs, references, controls,
        )
    }

    @JvmStatic
    @Throws(IOException::class)
    fun encodeReceipt(receipt: Receipt): ByteArray {
        val preservedPayloadDigests = receipt.preservedPayloadDigests.sorted()
        val protectionDigests = receipt.protectionDigests.sorted()
        return message { output ->
            output.writeString(1, receipt.schema)
            output.writeString(2, receipt.producer)
            output.writeString(3, receipt.project)
            output.writeString(4, receipt.variant)
            output.writeString(5, receipt.planSha256)
            output.writeString(6, receipt.inputAabSha256)
            output.writeString(7, receipt.outputAabSha256)
            output.writeString(8, receipt.resourceMappingSha256)
            output.writeUInt32(9, receipt.resourceIdCount)
            output.writeUInt32(10, receipt.renamedResourceCount)
            output.writeUInt32(11, receipt.renamedPathCount)
            output.writeBool(12, receipt.referenceClosureValidated)
            output.writeBool(13, receipt.preservedPayloadsValidated)
            output.writeBool(14, receipt.bundletoolValidated)
            output.writeUInt32(15, receipt.referenceCount)
            for (digest in preservedPayloadDigests) output.writeString(16, digest)
            for (digest in protectionDigests) output.writeString(17, digest)
        }
    }

    @Throws(IOException::class)
    private fun decodeResource(bytes: ByteArray): ResourceDecision {
        val input = CodedInputStream.newInstance(bytes)
        var id = 0
        val values = arrayOf("", "", "", "", "", "")
        val originalPaths = ArrayList<String>()
        val targetPaths = ArrayList<String>()
        var nameProtected = false
        var pathProtected = false
        while (!input.isAtEnd) {
            val tag = input.readTag()
            when (tag) {
                8 -> id = input.readUInt32()
                18, 26, 34, 42, 50, 58 ->
                    values[(tag - 18) / 8] = input.readStringRequireUtf8()
                66 -> originalPaths.add(input.readStringRequireUtf8())
                74 -> targetPaths.add(input.readStringRequireUtf8())
                80 -> nameProtected = input.readBool()
                88 -> pathProtected = input.readBool()
                0 -> { }
                else -> input.skipField(tag)
            }
        }
        return ResourceDecision(
            id, values[0], values[1], values[2], values[3],
            values[4], values[5], originalPaths, targetPaths,
            nameProtected, pathProtected,
        )
    }

    @Throws(IOException::class)
    private fun decodeEntry(bytes: ByteArray): EntryDecision {
        val input = CodedInputStream.newInstance(bytes)
        val values = arrayOf("", "", "", "")
        var method = 0
        var preserve = false
        while (!input.isAtEnd) {
            val tag = input.readTag()
            when (tag) {
                10 -> values[0] = input.readStringRequireUtf8()
                18 -> values[1] = input.readStringRequireUtf8()
                24 -> method = input.readUInt32()
                34 -> values[2] = input.readStringRequireUtf8()
                42 -> values[3] = input.readStringRequireUtf8()
                48 -> preserve = input.readBool()
                0 -> { }
                else -> input.skipField(tag)
            }
        }
        return EntryDecision(values[0], values[1], method, values[2], values[3], preserve)
    }

    @Throws(IOException::class)
    private fun decodeReference(bytes: ByteArray): ReferenceDecision {
        val input = CodedInputStream.newInstance(bytes)
        val values = arrayOf("", "", "", "", "")
        var id = 0
        while (!input.isAtEnd) {
            val tag = input.readTag()
            when (tag) {
                10, 18, 26 -> values[(tag - 10) / 8] = input.readStringRequireUtf8()
                32 -> id = input.readUInt32()
                42 -> values[3] = input.readStringRequireUtf8()
                50 -> values[4] = input.readStringRequireUtf8()
                0 -> { }
                else -> input.skipField(tag)
            }
        }
        return ReferenceDecision(values[0], values[1], values[2], id, values[3], values[4])
    }

    @Throws(IOException::class)
    private fun decodeControl(bytes: ByteArray): ControlDecision {
        val input = CodedInputStream.newInstance(bytes)
        val values = arrayOf("", "", "", "", "", "")
        var protectedTarget = false
        while (!input.isAtEnd) {
            val tag = input.readTag()
            when (tag) {
                10, 18, 26, 34, 42, 50 ->
                    values[(tag - 10) / 8] = input.readStringRequireUtf8()
                56 -> protectedTarget = input.readBool()
                0 -> { }
                else -> input.skipField(tag)
            }
        }
        return ControlDecision(
            values[0], values[1], values[2], values[3],
            values[4], values[5], protectedTarget,
        )
    }

    @Throws(IOException::class)
    private fun message(writer: (CodedOutputStream) -> Unit): ByteArray {
        val bytes = ByteArrayOutputStream()
        val output = CodedOutputStream.newInstance(bytes)
        writer(output)
        output.flush()
        return bytes.toByteArray()
    }

    private fun failure(
        project: String,
        variant: String,
        target: String,
        reason: String,
        repair: String,
    ): GradleException = KaleidoDiagnostic(
        "KLD-BUNDLE-001", project, variant, "bundle-rewrite",
        PLAN_SCHEMA, target, reason, repair,
    ).failure()

    @JvmRecord
    data class Plan private constructor(
        val schema: String,
        val producer: String,
        val project: String,
        val variant: String,
        val inputAabSha256: String,
        val resourceTableSha256: String,
        val resources: List<ResourceDecision>,
        val entries: List<EntryDecision>,
        val expectedOutputs: List<String>,
        val references: List<ReferenceDecision>,
        val controls: List<ControlDecision>,
        @Suppress("unused")
        private val canonical: Boolean,
    ) {
        constructor(
            schema: String,
            producer: String,
            project: String,
            variant: String,
            inputAabSha256: String,
            resourceTableSha256: String,
            resources: List<ResourceDecision>,
            entries: List<EntryDecision>,
            expectedOutputs: List<String>,
            references: List<ReferenceDecision>,
            controls: List<ControlDecision>,
        ) : this(
            schema,
            producer,
            project,
            variant,
            inputAabSha256,
            resourceTableSha256,
            resources.sortedWith { left, right -> Integer.compareUnsigned(left.id, right.id) },
            entries.sortedBy(EntryDecision::inputPath),
            expectedOutputs.sorted(),
            references.sortedWith(
                compareBy(ReferenceDecision::origin)
                    .thenBy(ReferenceDecision::fieldPath)
                    .thenBy(ReferenceDecision::kind),
            ),
            controls.sortedWith(
                compareBy(ControlDecision::kind).thenBy(ControlDecision::target),
            ),
            true,
        )
    }

    @JvmRecord
    data class ResourceDecision private constructor(
        val id: Int,
        val packageName: String,
        val type: String,
        val originalName: String,
        val targetName: String,
        val action: String,
        val reason: String,
        val originalPaths: List<String>,
        val targetPaths: List<String>,
        val nameProtected: Boolean,
        val pathProtected: Boolean,
        @Suppress("unused")
        private val canonical: Boolean,
    ) {
        constructor(
            id: Int,
            packageName: String,
            type: String,
            originalName: String,
            targetName: String,
            action: String,
            reason: String,
            originalPaths: List<String>,
            targetPaths: List<String>,
            nameProtected: Boolean,
            pathProtected: Boolean,
        ) : this(
            id,
            packageName,
            type,
            originalName,
            targetName,
            action,
            reason,
            canonicalOriginalPaths(originalPaths, targetPaths),
            canonicalTargetPaths(originalPaths, targetPaths),
            nameProtected,
            pathProtected,
            true,
        )
    }

    @JvmRecord
    data class EntryDecision(
        val inputPath: String,
        val inputSha256: String,
        val compressionMethod: Int,
        val outputPath: String,
        val action: String,
        val preservePayload: Boolean,
    )

    @JvmRecord
    data class ReferenceDecision(
        val origin: String,
        val fieldPath: String,
        val kind: String,
        val resourceId: Int,
        val originalValue: String,
        val targetValue: String,
    )

    @JvmRecord
    data class ControlDecision(
        val kind: String,
        val target: String,
        val inputSha256: String,
        val outputSha256: String,
        val action: String,
        val reason: String,
        val protectedTarget: Boolean,
    )

    @JvmRecord
    data class Receipt private constructor(
        val schema: String,
        val producer: String,
        val project: String,
        val variant: String,
        val planSha256: String,
        val inputAabSha256: String,
        val outputAabSha256: String,
        val resourceMappingSha256: String,
        val resourceIdCount: Int,
        val renamedResourceCount: Int,
        val renamedPathCount: Int,
        val referenceClosureValidated: Boolean,
        val preservedPayloadsValidated: Boolean,
        val bundletoolValidated: Boolean,
        val referenceCount: Int,
        val preservedPayloadDigests: List<String>,
        val protectionDigests: List<String>,
        @Suppress("unused")
        private val canonical: Boolean,
    ) {
        constructor(
            schema: String,
            producer: String,
            project: String,
            variant: String,
            planSha256: String,
            inputAabSha256: String,
            outputAabSha256: String,
            resourceMappingSha256: String,
            resourceIdCount: Int,
            renamedResourceCount: Int,
            renamedPathCount: Int,
            referenceClosureValidated: Boolean,
            preservedPayloadsValidated: Boolean,
            bundletoolValidated: Boolean,
            referenceCount: Int,
            preservedPayloadDigests: List<String>,
            protectionDigests: List<String>,
        ) : this(
            schema,
            producer,
            project,
            variant,
            planSha256,
            inputAabSha256,
            outputAabSha256,
            resourceMappingSha256,
            resourceIdCount,
            renamedResourceCount,
            renamedPathCount,
            referenceClosureValidated,
            preservedPayloadsValidated,
            bundletoolValidated,
            referenceCount,
            preservedPayloadDigests.sorted(),
            protectionDigests.sorted(),
            true,
        )
    }

    private fun canonicalOriginalPaths(
        originalPaths: List<String>,
        targetPaths: List<String>,
    ): List<String> {
        if (originalPaths.size != targetPaths.size) {
            throw IllegalArgumentException("Resource path mapping is not one-to-one")
        }
        return originalPaths.indices.sortedBy { originalPaths[it] }.map { originalPaths[it] }
    }

    private fun canonicalTargetPaths(
        originalPaths: List<String>,
        targetPaths: List<String>,
    ): List<String> {
        if (originalPaths.size != targetPaths.size) {
            throw IllegalArgumentException("Resource path mapping is not one-to-one")
        }
        return originalPaths.indices.sortedBy { originalPaths[it] }.map { targetPaths[it] }
    }
}
